core/src/main/java/org/jruby/javasupport/Java.java

```diff
@@ -430,7 +430,11 @@ public static RubyModule getProxyClass(Ruby runtime, JavaClass javaClass) {
     }
 
     @SuppressWarnings("deprecation")
-    public static RubyModule getProxyClass(final Ruby runtime, final Class<?> clazz) {
+    public static RubyModule getProxyClass(final Ruby runtime, Class<?> clazz) {
+        // skip synthetic classes we should not see, like InterfaceImpl*
+        while (clazz.isSynthetic()) {
```

Comment:
isn't there a better place to put to check, with regards to the interface impl that are problematic?

feels a bit confusing, and most callers do not need to worry about having a synthetic class here.

maybe add another `getClosestProxyClass` as an alternative? 
that way we would know which calls are internal and which are from user supplied classes...

Reply:
It turned out this is too aggressive; lambda expressions are synthetic and won't properly bind methods, and because we bind interface methods on the concrete type (interfaces don't really map to modules right) we always need that lowest concrete type to hold the bindings. But #8844 seems to fix it, even though it's much more invasive. We can hide it behind an option for users that run into this, and we should also finally fix the re-use of interface impl classes.